src/puffin/src/file_format/writer/file.rs

```diff
@@ -0,0 +1,156 @@
+use futures::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use snafu::ResultExt;
+
+use crate::blob_metadata::{BlobMetadata, BlobMetadataBuilder};
+use crate::error::{Result, WriteSnafu};
+use crate::file_format::writer::footer::FooterWriter;
+use crate::file_format::writer::{Blob, PuffinAsyncWriter, PuffinSyncWriter};
+use crate::file_format::MAGIC;
+
+/// Puffin file writer, implements both [`PuffinSyncWriter`] and [`PuffinAsyncWriter`]
+pub struct PuffinFileWriter<W> {
+    /// The writer to write to
+    writer: W,
+
+    /// The properties of the file
+    properties: HashMap<String, String>,
+
+    /// The metadata of the blobs
+    blob_metadata: Vec<BlobMetadata>,
+
+    /// The offset of the next blob
+    next_blob_offset: u64,
+}
+
+impl<W> PuffinFileWriter<W> {
+    pub fn new(writer: W) -> Self {
+        Self {
+            writer,
+            properties: HashMap::new(),
+            blob_metadata: Vec::new(),
+            next_blob_offset: 0,
+        }
+    }
+
+    fn create_blob_metadata(
+        &self,
+        typ: String,
+        properties: HashMap<String, String>,
+        size: u64,
+    ) -> BlobMetadata {
+        BlobMetadataBuilder::default()
+            .blob_type(typ)
+            .properties(properties)
+            .offset(self.next_blob_offset as _)
+            .length(size as _)
+            .build()
+            .expect("Required fields are not set")
+    }
+}
+
+impl<W: io::Write> PuffinSyncWriter for PuffinFileWriter<W> {
+    fn set_properties(&mut self, properties: HashMap<String, String>) {
+        self.properties = properties;
+    }
+
+    fn add_blob<R: io::Read>(&mut self, mut blob: Blob<R>) -> Result<()> {
+        self.write_header_if_needed_sync()?;
+
+        let size = io::copy(&mut blob.data, &mut self.writer).context(WriteSnafu)?;
+
+        let blob_metadata = self.create_blob_metadata(blob.blob_type, blob.properties, size);
+        self.blob_metadata.push(blob_metadata);
+
+        self.next_blob_offset += size;
+        Ok(())
+    }
+
+    fn finish(&mut self) -> Result<()> {
+        self.write_header_if_needed_sync()?;
+        self.write_footer_sync()
+    }
+}
+
+#[async_trait]
+impl<W: AsyncWrite + Unpin + Send> PuffinAsyncWriter for PuffinFileWriter<W> {
+    fn set_properties(&mut self, properties: HashMap<String, String>) {
+        self.properties = properties;
+    }
+
+    async fn add_blob<R: AsyncRead + Send>(&mut self, blob: Blob<R>) -> Result<()> {
+        self.write_header_if_needed_async().await?;
+
+        let size = futures::io::copy(blob.data, &mut self.writer)
+            .await
+            .context(WriteSnafu)?;
+
+        let blob_metadata = self.create_blob_metadata(blob.blob_type, blob.properties, size);
+        self.blob_metadata.push(blob_metadata);
+
+        self.next_blob_offset += size;
+        Ok(())
+    }
+
+    async fn finish(&mut self) -> Result<()> {
+        self.write_header_if_needed_async().await?;
+        self.write_footer_async().await
```

Comment:
We might need to call `flush()` explicitly to ensure data is flushed, especially for async writers as they can't flush on drop.

Reply:
updated